sympy/polys/rootoftools.py

```diff
@@ -772,15 +775,64 @@ def _get_roots(cls, method, poly, radicals):
         free_names = {str(i) for i in poly.free_symbols}
         for x in chain((symbols('x'),), numbered_symbols('x')):
             if x.name not in free_names:
-                poly = poly.xreplace({d: x})
+                poly = poly.replace(d, x)
                 break
+
+        if dom.is_QQ or dom.is_ZZ:
+            return cls._get_roots_qq(method, poly, radicals)
+        elif dom.is_AlgebraicField or dom.is_ZZ_I or dom.is_QQ_I:
+            return cls._get_roots_alg(method, poly, radicals)
+        else:
+            # XXX: not sure how to handle ZZ[x] which appears in some tests?
+            # this makes the tests pass alright but has to be a better way?
+            return cls._get_roots_qq(method, poly, radicals)
+
+
+    @classmethod
+    def _get_roots_qq(cls, method, poly, radicals):
+        """Return postprocessed roots of specified kind
+         for polynomials with rational coefficients. """
         coeff, poly = cls._preprocess_roots(poly)
         roots = []
 
         for root in getattr(cls, method)(poly):
             roots.append(coeff*cls._postprocess_root(root, radicals))
+
         return roots
 
+    @classmethod
+    def _get_roots_alg(cls, method, poly, radicals):
+        """Return postprocessed roots of specified kind
+         for polynomials with algebraic coefficients. It assumes
+         the domain is already an algebraic field. First it
+         finds the roots using _get_roots_qq, then uses the
+         square-free factors to filter roots and get the correct
+         multiplicity.
+         """
+
+        # Existing QQ code can find and sort the roots
+        roots = cls._get_roots_qq(method, poly.lift(), radicals)
+
+        subroots = {}
+        for f, m in poly.sqf_list()[1]:
+            if method == "_real_roots":
+                roots_filt = f.which_roots(roots, real=True)
+            elif method == "_all_roots":
+                roots_filt = f.which_roots(roots, real=False)
+            for r in roots_filt:
+                if r in subroots:
+                    subroots[r] += m
+                else:
+                    subroots[r] = m
```

Comment:
We can write this as just `subroots[r] = m`.

Is it possible for distinct square-free factors to have any roots in common?

I don't think it is. Let's suppose that `fi` and `fj` are factors with multiplicity `mi` and `mj`. If some number `r` is a root of both `fi` and `fj` then it must be a root of some `fij` that is a factor of both `fi` and `fj` and then the multiplicity of `fij` would be at least `mi + mj`.

To put it another way the distinct square-free factorisation is defined so that the factors are pairwise coprime (the algorithm literally divides out the gcds). That means that they cannot have any roots in common.

Reply:
Ah great point, I put that in case, but from your explanation I can see it wouldn't happen.